File: dotnet/spa/PkiSuiteAspNetSpaSample/Controllers/XmlNFeSignatureRestController.cs
using Lacuna.RestPki.Api;
using Lacuna.RestPki.Client;
using Microsoft.AspNetCore.Mvc;
using PkiSuiteAspNetSpaSample.Classes;
using PkiSuiteAspNetSpaSample.Models.Rest;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace PkiSuiteAspNetSpaSample.Controllers {
	[Route("api/[controller]/[action]")]
	[ApiController]
	public class XmlNFeSignatureRestController : ControllerBase {
		private readonly StorageMock _storageMock;
		private readonly Util _util;

		public XmlNFeSignatureRestController(StorageMock storageMock, Util util)
		{
			_storageMock = storageMock;
			_util = util;
		}

		/**
		 * GET: XmlNFeSignatureRest
		 * 
		 * This action is called once the user's certificate encoding has been read, and contains the
		 * logic to prepare the hash that needs to be actually signed with the user's private key
		 * (the "to-sign-hash").
		 */
		[HttpGet]
		public async Task<Models.Rest.SignatureStartResponse> StartAsync()
		{

			// Instantiate the XmlElementSignatureStarter class, responsible for receiving the signature
			// elements and start the signature process.
			var signatureStarter = new XmlElementSignatureStarter(_util.GetRestPkiClient());

			// Set the XML to be signed, a sample Brazilian fiscal invoice pre-generated.
			signatureStarter.SetXml(_storageMock.GetSampleNFePath());

			// Set the ID of the element to be signed.
			signatureStarter.SetToSignElementId("NFe35141214314050000662550010001084271182362300");

			// Set the signature policy.
			signatureStarter.SetSignaturePolicy(StandardXmlSignaturePolicies.PkiBrazil.NFePadraoNacional);

			// Set the security context to be used to determine trust in the certificate chain. We have
			// encapsulated the security context choice on Util.cs.
			signatureStarter.SetSecurityContext(Util.GetSecurityContextId());

			// Call the StartWithWebPki() method, which initiates the signature. This yields the token,
			// a 43-character case-sensitive URL-safe string, which identifies this signature process. We'll
			// use this value to call the signWithRestPki() method on the Web PKI component (see
			// signature-form.js) and also to complete the signature on the POST action below (this should
			// not be mistaken with the API access token).
			var token = await signatureStarter.StartWithWebPkiAsync();
			return new Models.Rest.SignatureStartResponse()
			{
				Token = token
			};
		}

		/**
		 * POST: XmlNFeSignatureRest/Complete
		 * 
		 * This action is called once the "to-sign-hash" are signed using the user's 
		 * certificate. After signature, it'll return the signature file.
		 */
		[HttpPost]
		public async Task<SignatureCompleteResponse> CompleteAsync(SignatureCompleteRequest request)
		{
			// Get an instance of the XmlSignatureFinisher class, responsible for completing the signature
			// process.
			var signatureFinisher = new XmlSignatureFinisher(_util.GetRestPkiClient())
			{
				// Set the token for this signature (rendered in a hidden input field, see the view).
				Token = request.Token
			};

			// Call the Finish() method, which finalizes the signature process and returns the signed PDF.
			var signedXml = await signatureFinisher.FinishAsync();

			// Optional - Get information about the signer's certificate used.
			// This method must only be called after calling the Finish() method.
			//var signerCert = signatureFinisher.GetCertificateInfo();

			// At this point, you'd typically store the signed XML on your database. For demonstration
			// purposes, we'll store the PDF on our mock Storage class.
			var fileId = _storageMock.Store(signedXml, ".xml");

			return new SignatureCompleteResponse()
			{
				SignedFileId = fileId,
			};
		}
	}
}

File: java/springmvc/src/main/java/com/lacunasoftware/pkisuite/controller/XmlNFeSignatureRestController.java
package com.lacunasoftware.pkisuite.controller;

import com.lacunasoftware.pkisuite.util.SampleDocs;
import com.lacunasoftware.pkisuite.util.StorageMock;
import com.lacunasoftware.pkisuite.util.Util;
import com.lacunasoftware.restpki.*;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;


@Controller
@RequestMapping("/xml-nfe-signature-rest")
public class XmlNFeSignatureRestController {

	/**
	 * GET /xml-nfe-signature-rest
	 *
	 * This route initiates a XML element signature using REST PKI and 
	 * renders the signature page.
	 * 
	 * The XML element signature is recommended in cases which there is
	 * a need to sign a specific element of a XML.
	 */
	@GetMapping
	public String get(
		Model model,
		HttpServletResponse response
	) throws IOException, RestException {
		// Instantiate the XmlElementSignatureStarter class, responsible for receiving the signature
		// elements and start the signature process.
		XmlElementSignatureStarter signatureStarter = new XmlElementSignatureStarter(Util.getRestPkiClient());

		// Set the XML to be signed, a sample Brazilian fiscal invoice pre-generated.
		signatureStarter.setXml(StorageMock.getSampleDocPath(SampleDocs.SAMPLE_NFE));

		// Set the ID of the element to be signed.
		signatureStarter.setElementToSIgnId("NFe35141214314050000662550010001084271182362300");

		// Set the signature policy.
		signatureStarter.setSignaturePolicy(SignaturePolicy.NFePadraoNacional);

		// Set the security context to be used to determine trust in the certificate chain. We have
		// encapsulated the security context choice on Util.java.
		signatureStarter.setSecurityContext(Util.getSecurityContext());

		// Call the startWithWebPki() method, which initiates the signature. This yields the token,
		// a 43-character case-sensitive URL-safe string, which identifies this signature process.
		// We'll use this value to call the signWithRestPki() method on the Web PKI component
		// (see file signature-form.js) and also to complete the signature after the form is
		// submitted (see method post() below). This should not be mistaken with the API access
		// token.
		String token = signatureStarter.startWithWebPki();

		// The token acquired above can only be used for a single signature attempt. In order to
		// retry the signature it is necessary to get a new token. This can be a problem if the user
		// uses the back button of the browser, since the browser might show a cached page that we
		// rendered previously, with a now stale token. To prevent this from happening, we call the
		// method Util.setNoCacheHeaders(), which sets HTTP headers to prevent caching of the page.
		Util.setNoCacheHeaders(response);

		// Render the signature page (templates/xml-diploma-signature-rest.html).
		model.addAttribute("sampleId", SampleDocs.SAMPLE_NFE);
		model.addAttribute("token", token);
		return "xml-nfe-signature-rest/index";
	}

	/**
	 * POST /xml-nfe-signature-rest
	 *
	 * This route receives the form submission from the view 
	 * 'xml-nfe-signature'. We'll call REST PKI to complete the signature.
	 */
	@PostMapping
	public String post(@RequestParam String token, Model model) throws IOException, RestException {
		// Instantiate the XmlSignatureFinisher class, responsible for completing the signature
		// process.
		XmlSignatureFinisher signatureFinisher = new XmlSignatureFinisher(Util.getRestPkiClient());

		// Set the token for this signature (rendered in a hidden input field, see
		// file templates/xml-full-signature.html).
		signatureFinisher.setToken(token);

		// Call the finish() method, which finalizes the signature process and returns the signed
		// XML's bytes.
		byte[] signedXml = signatureFinisher.finish();

		// Get information about the certificate used by the user to sign the file. This method must
		// only be called after calling the finish() method.
		PKCertificate signerCert = signatureFinisher.getCertificateInfo();

		// At this point, you'd typically store the signed PDF on your database. For demonstration
		// purposes, we'll store the XML on a temporary folder and return to the page an identifier
		// that can be used to download it.
		String filename = StorageMock.store(signedXml, "xml");

		// Render the signature page (templates/xml-signature-info.html).
		model.addAttribute("signerCert", signerCert);
		model.addAttribute("filename", filename);
		return "xml-signature-rest/signature-info";
	}
}
